### ISSUE TITLE
faceswap video result bad after 120k training using FaceSwap_GAN_v2_train, but the preview is not bad

### ISSUE THREAD

**Issue by hengshanji** [shaoanlu/faceswap-GAN#83]
Firstly thank you for your detailed work , i am very interested in the faceswap. But i have not much knowledge about this.  After one week , my model is still not well.  This is previews after 12w train:
![100000_mask](https://user-images.githubusercontent.com/7146801/37320768-36b8230a-26b0-11e8-9d8b-9806e181b90c.jpg)
![100000_masked](https://user-images.githubusercontent.com/7146801/37320769-36f44808-26b0-11e8-92c4-cebae2d39d27.jpg)

this is video using models:

![trump_output_v2_12w](https://user-images.githubusercontent.com/7146801/37320970-3bc65726-26b1-11e8-8186-8fa8b5b5ee9d.gif)

My dataset has about 300 picture for each person, the pictures are from different lighting, position and so on, they are from different video.
I have read almost all the issues. The video result is like issues/50 but the reason is on the contrary. Is the possible my dataset has too much diversity and this could make the model very unstable ?  I also make the roi size bigger because i find the jaw is very blurred sometime and this works indeed. However there are still frames which is very unreal, i save the origin roi patch and the fake roi patch:
![00001-c](https://user-images.githubusercontent.com/7146801/37321054-a0be97a6-26b1-11e8-904e-a98d29b5316a.jpg)
![00002](https://user-images.githubusercontent.com/7146801/37321056-a11a0302-26b1-11e8-8c1f-03ea0f709f72.jpg)
![00002-c](https://user-images.githubusercontent.com/7146801/37321057-a15e2348-26b1-11e8-8180-0ef165bb9dda.jpg)
![00001](https://user-images.githubusercontent.com/7146801/37321059-a19f1b32-26b1-11e8-9a87-0f71c83c8b30.jpg)

the two frames above are good 

![00029](https://user-images.githubusercontent.com/7146801/37321100-c9989938-26b1-11e8-9e7a-8034692a4c89.jpg)
![00029-c](https://user-images.githubusercontent.com/7146801/37321101-c9df78da-26b1-11e8-9e5e-3fe1f97ef71c.jpg)
![00030](https://user-images.githubusercontent.com/7146801/37321103-ca31a3ee-26b1-11e8-92b7-a362df22a8c5.jpg)
![00030-c](https://user-images.githubusercontent.com/7146801/37321104-ca728e4a-26b1-11e8-861a-9640f7f09088.jpg)
![00033](https://user-images.githubusercontent.com/7146801/37321105-cac39fec-26b1-11e8-8023-d338a994328d.jpg)
![00033-c](https://user-images.githubusercontent.com/7146801/37321107-cb1017a0-26b1-11e8-813f-c0e0e28453cb.jpg)
![00005-c](https://user-images.githubusercontent.com/7146801/37321108-cb47f0ee-26b1-11e8-9f38-2b56f16eb9a0.jpg)

these are bad results. It seems the mouth and nose part are not swapped.

Now i am retraining the model using three method,  first is using perceptual loss with v2_train, second is using v2_sz128_train with perceptual loss, third is add training pictures to 1k for each person.

Could you please help me to analyze the reason why the converted video is so unstable and blurred?  I will be very appreciated!!

**Comment by shaoanlu** [comment]
From the previews you provided, I assume the test video is not included in the training data. In this case, 12w (120k?) iters are too much that the model is heavily overfit. In my opinion, 12k iters (6000 iters without perceptual loss following by 6000 iters with perceptual loss) is enough.

**Comment by slundell** [comment]
When you say 6k + 6k iterations are enough, you mean iterations of one batch, and not epochs? What batch size are you using as that will affect the number of iterations needed

**Comment by hengshanji** [comment]
@shaoanlu  thank you so much, i will try it!

**Comment by hengshanji** [comment]
I have train the model as your comment  6k+6k. The mask looks like good ,but the part of eye and nose in the preview image is blurry.
This is the previews 12k
![12000_masked](https://user-images.githubusercontent.com/7146801/37473968-40d63448-28aa-11e8-9b4a-4c21b509f80f.jpg)
![10000_mask](https://user-images.githubusercontent.com/7146801/37474036-637c91c2-28aa-11e8-8171-946c973bccaf.jpg)

The converted video is:
<img width="935" alt="v3_test" src="https://user-images.githubusercontent.com/7146801/37476042-17393504-28af-11e8-9cd5-005d73746101.png">

So do i need add more pictures to retrain the model ?  or should i use other people's picture to retrain the model?

**Comment by shaoanlu** [comment]
It seems to me that PL was not introduced during training. Can you describe the steps you took to activate PL?

**Comment by hengshanji** [comment]
I change the use_perceptual_loss to True to activate PL in FaceSwap_GAN_v2_train.ipynb. Is that not right?

**Comment by shaoanlu** [comment]
Once the `use_perceptual_loss` is set to True, Cells in step 8, Define Loss Function,  should be run once again.

**Comment by hengshanji** [comment]
Actually i use jupyter nbconvert  change the ipynb to py script. I first train 6k and stop the scipt , change use_perceptual_loss  to True and then i run the script again.

**Comment by shaoanlu** [comment]
These are the I got output after 6900+1500 iters training. I made some modifications in the perceptual loss and the masks now make sense to me. 
**Masked**
![8400masked](https://user-images.githubusercontent.com/22006284/37511384-921e4c48-2939-11e8-98e8-9a6388291d1f.png)

**Raw**
![8400raw](https://user-images.githubusercontent.com/22006284/37511401-a2acd35e-2939-11e8-9c87-246cd84c0632.png)

**Mask**
![8400mask](https://user-images.githubusercontent.com/22006284/37511406-a8027b38-2939-11e8-8661-817dc724cea7.png)

In step 8, Define Loss Functions `def define_loss(...)`, the original code
```python
        ...
        pl_params = (0.011, 0.11, 0.1919)
        real_sz224 = tf.image.resize_images(real, [224, 224])
        real_sz224 = Lambda(preprocess_vggface)(real_sz224)
        # ==========
        fake_sz224 = tf.image.resize_images(fake_rgb, [224, 224]) 
        fake_sz224 = Lambda(preprocess_vggface)(fake_sz224)
        # ==========   
        real_feat55, real_feat28, real_feat7 = vggface_feat(real_sz224)
        fake_feat55, fake_feat28, fake_feat7  = vggface_feat(fake_sz224)    
        loss_G += pl_params[0] * K.mean(K.abs(fake_feat7 - real_feat7))
        loss_G += pl_params[1] * K.mean(K.abs(fake_feat28 - real_feat28))
        loss_G += pl_params[2] * K.mean(K.abs(fake_feat55 - real_feat55))
    
    return loss_D, loss_G
```
is changed as
```python
        ...
        pl_params = (0.03, 0.2, 0.3) # increase weight
        real_sz224 = tf.image.resize_images(real, [224, 224])
        real_sz224 = Lambda(preprocess_vggface)(real_sz224)
        # ==========
        fake_sz224 = tf.image.resize_images(fake, [224, 224]) 
        fake_sz224 = Lambda(preprocess_vggface)(fake_sz224)
        # ==========   
        real_feat55, real_feat28, real_feat7 = vggface_feat(real_sz224)
        fake_feat55, fake_feat28, fake_feat7  = vggface_feat(fake_sz224)    
        loss_G += pl_params[0] * K.mean(K.abs(fake_feat7 - real_feat7))
        loss_G += pl_params[1] * K.mean(K.abs(fake_feat28 - real_feat28))
        loss_G += pl_params[2] * K.mean(K.abs(fake_feat55 - real_feat55))
        
        # Additional perceptual loss to raw output
        pl_params = (0.011, 0.11, 0.1919)
        # ==========
        fake_sz224 = tf.image.resize_images(fake_rgb, [224, 224]) 
        fake_sz224 = Lambda(preprocess_vggface)(fake_sz224)
        # ==========   
        fake_feat55, fake_feat28, fake_feat7  = vggface_feat(fake_sz224)    
        loss_G += pl_params[0] * K.mean(K.abs(fake_feat7 - real_feat7))
        loss_G += pl_params[1] * K.mean(K.abs(fake_feat28 - real_feat28))
        loss_G += pl_params[2] * K.mean(K.abs(fake_feat55 - real_feat55))
    
    return loss_D, loss_G
```

Besides, I also zoom-in the training images by changing params of `def random_warp( image )` in `image_augmentation.py` as the following:
```python
range_ = numpy.linspace( 128-80, 128+80, 5 )
#range_ = numpy.linspace( 128-110, 128+110, 5 )
```

**Comment by shaoanlu** [comment]
To check if PL is activated, you can compare the training time per iter before/after activating PL. The latter should be 1.5x ~2x longer.
